helpers/kernel_symbols.go

```diff
@@ -3,340 +3,315 @@ package helpers
+	k.updateLock.RLock()
+	defer k.updateLock.RUnlock()
+
+	symbols, exist := k.addrs[addr{a}]
+	if !exist {
+		return []KernelSymbol{}, symNotFoundErr(a)
 	}
-	return nil, SymbolNotFound(owner, name)
+
+	return copySliceOfPointersToSliceOfStructs(symbols), nil
 }
 
-// GetSymbolByAddr returns a symbol by a given address
-func (k *fullKernelSymbolTable) GetSymbolByAddr(addr uint64) (*KernelSymbol, error) {
-	symbol, exist := k.symbolAddrMap[addr]
-	if exist {
-		return symbol, nil
+// GetSymbolByOwnerAndName returns all the symbols with the given owner and name.
+func (k *KernelSymbolTable) GetSymbolByOwnerAndName(o, n string) ([]KernelSymbol, error) {
+	k.updateLock.RLock()
+	defer k.updateLock.RUnlock()
+
+	symbols, exist := k.symByName[nameAndOwner{n, o}]
+	if !exist {
+		return []KernelSymbol{}, symNotFoundErr(nameAndOwner{n, o})
 	}
-	return nil, SymbolNotFoundAtAddress(addr)
+
+	return copySliceOfPointersToSliceOfStructs(symbols), nil
 }
 
-func (k *fullKernelSymbolTable) Refresh() error {
-	k.symbolMap = make(map[string]*KernelSymbol)
-	k.symbolAddrMap = make(map[uint64]*KernelSymbol)
-	file, err := os.Open(kallsymsPath)
-	if err != nil {
-		return fmt.Errorf("could not open /proc/kallsyms: %w", err)
+// GetSymbolByOwnerAndAddr returns all the symbols with the given owner and address.
+func (k *KernelSymbolTable) GetSymbolByOwnerAndAddr(o string, a uint64) ([]KernelSymbol, error) {
+	k.updateLock.RLock()
+	defer k.updateLock.RUnlock()
+
+	symbols, exist := k.symByAddr[addrAndOwner{a, o}]
+	if !exist {
+		return []KernelSymbol{}, symNotFoundErr(addrAndOwner{a, o})
 	}
-	defer file.Close()
-	scanner := bufio.NewScanner(file)
-	scanner.Split(bufio.ScanLines)
-	for scanner.Scan() {
-		line := strings.Fields(scanner.Text())
-		// if the line is less than 3 words, we can't parse it (one or more
-		// fields missing)
-		if len(line) < 3 {
-			continue
-		}
-		symbolAddr, err := strconv.ParseUint(line[0], 16, 64)
-		if err != nil {
-			continue
-		}
 
-		symbolType, symbolName, symbolOwner := parseSymbolLine(line)
+	return copySliceOfPointersToSliceOfStructs(symbols), nil
+}
 
-		symbolKey := symbolOwner + "_" + symbolName
-		symbol := &KernelSymbol{symbolName, symbolType, symbolAddr, symbolOwner}
-		k.symbolMap[symbolKey] = symbol
-		k.symbolAddrMap[symbolAddr] = symbol
-	}
-	stext, err := k.GetSymbolByName("system", "_stext")
-	if err != nil {
-		return err
-	}
-	k.textSegStart = stext.Address
-	etext, err := k.GetSymbolByName("system", "_etext")
+// Concurrency logic for updating the maps: faster than a single goroutine that
+// updates all maps OR multiple goroutines with fine grained locking (turned out
+// to be slower than a single goroutine).
+//
+// Simple file parsing (no processing) takes ~0.200 seconds on a 4-core machine.
+// If buffer is increased to 4MB, it might take ~0.150 seconds. A simple
+// mono-threaded implementation that parses + processes the lines takes ~0.700
+// seconds. This approach takes ~0.350 seconds (2x speedup).
```

Comment:
I'm not doing performance tuning/benchmarking for this change (since Im actually removing one attempt to do so that caused the original problem because of wrong premises). Be my guest to open an issue and benchmark any optimization to this code =).